### PR TITLE
LibWasm+wasm: Start implementing WASI

### PR DESCRIPTION
**Pull request by alimpfard**

### CODE REVIEW

File: Userland/Libraries/LibWasm/WASI/Wasi.cpp
Comment (LucasChollet) [inline Userland/Libraries/LibWasm/WASI/Wasi.cpp:156]: Commented-out code :scream:

Comment (alimpfard) [reply Userland/Libraries/LibWasm/WASI/Wasi.cpp]: :scream:

File: Userland/Utilities/wasm.cpp
Comment (ADKaster) [inline Userland/Utilities/wasm.cpp:390]: does LibFilesystem not have this API yet? It would be preferred to use that over adding more uses of DeprecatedFile.

**Review by ADKaster — COMMENTED** [review]
Looks like a decent framework to start! I don't really understand the serialization code though. it feels awkward. Like perhaps it should be using writeable spans rather than Rvalue refs to arrays?

File: Userland/Libraries/LibWasm/Wasi.h
Comment (alimpfard) [reply Userland/Libraries/LibWasm/Wasi.h:26]: Ah...I'll clean that up :P

File: Userland/Libraries/LibWasm/Wasi.h
Comment (alimpfard) [reply Userland/Libraries/LibWasm/Wasi.h:87]: We're serialising into a span, who owns the span is really not relevant.
Ideally I'd have just called it `serialize_into(Bytes)`, but there are a couple return types that take two spans, so we have to either take these as variadic args (which are a usability nightmare), or an `Array<Bytes>`
I can make this `Array<Bytes, 1> bytes`, but there's no real difference aside from copying the Bytes objects every time :shrug:

File: Userland/Libraries/LibWasm/Wasi.h
Comment (alimpfard) [reply Userland/Libraries/LibWasm/Wasi.h:120]: They are: https://github.com/WebAssembly/wasi-libc/blob/2c2fc9a2fddd0927a66f1c142e65c8dab6f5c5d7/libc-bottom-half/headers/public/wasi/api.h#L110

File: Userland/Libraries/LibWasm/WASI/Wasi.cpp
Comment (alimpfard) [reply Userland/Libraries/LibWasm/WASI/Wasi.cpp:36]: Into the span pointed by the Bytes - those are spans, not ByteBuffers.

File: Userland/Libraries/LibWasm/WASI/Wasi.cpp
Comment (alimpfard) [reply Userland/Libraries/LibWasm/WASI/Wasi.cpp:679]: did I add that? I think I just copied that from some random file :thinking:

File: Userland/Libraries/LibWasm/Wasi.h
Comment (alimpfard) [reply Userland/Libraries/LibWasm/Wasi.h:238]: :thinking: we don't really ever access those bits, but if it makes you (or the reader) feel better, sure.

File: Userland/Libraries/LibWasm/Wasi.h
Comment (alimpfard) [reply Userland/Libraries/LibWasm/Wasi.h:499]: Yep: <https://github.com/WebAssembly/wasi-libc/blob/2c2fc9a2fddd0927a66f1c142e65c8dab6f5c5d7/libc-bottom-half/headers/public/wasi/api.h#L1348>

**Comment by alimpfard** [comment]
> Like perhaps it should be using writeable spans rather than Rvalue refs to arrays

`Bytes` is _the_ writeable `Span<u8>`, the array is the "s" in spans - I can do two other things:
- Take variadic `Bytes` args (pls no)
- Take `Array<Bytes, N>` (pointless copy) or `Array<Bytes, N>&` (extra useless variable)

File: Userland/Utilities/wasm.cpp
Comment (alimpfard) [reply Userland/Utilities/wasm.cpp:390]: LibFileSystem's APIs use the new String, LexicalPath and Core::System use DeprecatedString - going through LibFileSystem is gonna make us do DeprecatedString -> String -> DeprecatedString.
I'd rather just keep the DeprecatedFile and not pretend this is using DeprecatedStuff.

File: Userland/Libraries/LibWasm/Wasi.h
Comment (alimpfard) [reply Userland/Libraries/LibWasm/Wasi.h]: made it take `Array<Bytes, N>`.

File: Userland/Libraries/LibWasm/Wasi.h
Comment (alimpfard) [reply Userland/Libraries/LibWasm/Wasi.h]: Added a comment for future me :shrug:

File: Userland/Libraries/LibWasm/WASI/Wasi.cpp
Comment (alimpfard) [reply Userland/Libraries/LibWasm/WASI/Wasi.cpp]: `#pragma`'d

**Review by ADKaster — APPROVED** [review]
No qualms with this on my end, but I'll let you hit the merge button in case you think it should go through another CI run post-rebase.

**Comment by alimpfard** [comment]
CI error seems irrelevant <https://dev.azure.com/SerenityOS/SerenityOS/_build/results?buildId=30793&view=logs&j=2351e412-c793-598f-27f8-72c4f2d55ba1&t=ee375dfa-6b7d-514a-7789-3948d04c0dd3&l=357>

**Pull request merged**
